runtime/src/main/java/org/corfudb/runtime/collections/streaming/StreamPollingScheduler.java

```diff
@@ -0,0 +1,309 @@
+    }
+
+    private Interval merge(Interval interval, StreamAddressRange range) {
+        long newMinX = Math.min(interval.x, range.getEnd());
+        long newMaxY = Math.max(interval.y, range.getStart());
+        return new Interval(newMinX, newMaxY);
+    }
+
+    // Generate a range per stream that contains all sub-ranges that need to be queries
+    private List<StreamAddressRange> coalesce(List<StreamAddressRange> ranges) {
+        Map<UUID, Interval> adjusted = new HashMap<>();
+
+        for (StreamAddressRange range : ranges) {
+            // (end, start]
+            Interval interval = adjusted.getOrDefault(range.getStreamID(), new Interval(Long.MAX_VALUE, Long.MIN_VALUE));
+            Interval expandedInterval = merge(interval, range);
+            Preconditions.checkState(expandedInterval.x >= Address.NEVER_READ);
+            Preconditions.checkState(expandedInterval.x < expandedInterval.y);
+
+            adjusted.put(range.getStreamID(), expandedInterval);
+        }
+
+        List<StreamAddressRange> adjustedList = new ArrayList<>();
+        for (Map.Entry<UUID, Interval> entry : adjusted.entrySet()) {
+            adjustedList.add(new StreamAddressRange(entry.getKey(), entry.getValue().y, entry.getValue().x));
+        }
+        return adjustedList;
+    }
+
+
+    // send notification internal queing consumes cpu even if the poll returns no results
+
+    private void dispatchSyncTasks(List<StreamingTask> tasks) {
+        for (StreamingTask task : tasks) {
+            if (task.getStream().hasNext()) {
+                // this will fail for tasks that are already syncing
+                // for tasks to be scheduled to sync, they must be in the scheduling state
+                task.move(StreamStatus.SCHEDULING, StreamStatus.SYNCING);
+                workers.execute(task);
+            } else {
+                // no new deltas to sync
+                // This stream has no entries to sync after the poll, set it to runnable so that it can
+                // be polled next scheduling cycle
+                task.move(StreamStatus.SCHEDULING, StreamStatus.RUNNABLE);
+            }
+        }
+    }
+
+    private void handleFailures(List<StreamingTask> failedTasks) {
+        // Need to remove the tasks from allTasks before propagating the error just incase they try to
+        // re-register the task?
+        failedTasks.forEach(t -> allTasks.remove(t.getListener()));
+
+        for (StreamingTask task : failedTasks) {
+            workers.execute(task::propagateError);
+        }
+    }
+
+    // hide this?
+    public void schedule() throws InterruptedException {
+        // Block thread, so that it doesn;t spin if there are no tasks
+        waitForTasks();
+
+        long startTime = System.nanoTime();
+
+        List<StreamingTask> runnableTasks = getTasks(StreamStatus.RUNNABLE);
+        List<StreamingTask> syncingTasks = getTasks(StreamStatus.SYNCING, t -> t.getStream().availableSpace() >= pollThreshold);
```

Comment:
Since a task can change from SYNCING to RUNNABLE while the scheduling is running, its imperative that the RUNNABLE tasks are queried first, otherwise there could be common tasks in runnableTasks and syncingTasks, this can cause issues if the same task is dispatched to sync on two different worker threads.